### PR TITLE
BUG: Correct datetime64 missing type overload for datetime.date #18640

### LINKED ISSUE #18640: datetime64 missing type overload for datetime.date
### Typing code example:

```python
import datetime
import numpy as np
d = np.datetime64(datetime.date(2021, 3, 17))
```

### Run mypy on file

```
untitled.py:3: error: No overload variant of "datetime64" matches argument type "date"  [call-overload]
untitled.py:3: note: Possible overload variant:
untitled.py:3: note:     def datetime64(self, Union[None, datetime64, Union[str, bytes], datetime] = ..., Union[Union[str, bytes], Tuple[Union[str, bytes], Union[int, integer[Any]]]] = ...) -> datetime64
scripts\untitled0.py:3: note:     <1 more non-matching overload not shown>
Found 1 error in 1 file (checked 1 source file)
```

Running the code with a datetime.datetime works fine, so it appears that numpy is missing datetime.date in the Union of allowable types to pass to np.datetime64

### NumPy/Python version information:

1.20.1 3.9.2 (tags/v3.9.2:1a79785, Feb 19 2021, 13:44:55) [MSC v.1928 64 bit (AMD64)]

### PR DESCRIPTION
**Pull request by oscar139**

### CODE REVIEW

File: numpy/__init__.pyi [hunk 1]
```diff
@@ -2537,11 +2537,24 @@ class object_(generic):
 
 object0 = object_
 
+#Added Protocol to typecheck for day, month, year property to be used in
+#datetime64 constructor per Issue #18640 and BvB93s recommendation
```
Comment (BvB93) [inline numpy/__init__.pyi:2541]:
```suggestion
# The `datetime64` constructors requires an object with the three attributes below, 
# and thus supports datetime duck typing
```
I feel that this comment would describe a bit more concisely what exactly is going on here.

File: numpy/__init__.pyi [hunk 2]
```diff
@@ -2537,11 +2537,24 @@ class object_(generic):
...
+    @property
+    def month(self) -> int: ...
+    @property
+    def year(self) -> int: ...
+
+
 class datetime64(generic):
-    @overload
+    #Replaced dt.datetime with _DatetimeScalar per issue #18640 and
+    #BvB93s recommendation
```
Comment (BvB93) [inline numpy/__init__.pyi:2553]:
```suggestion
```
I'd actually suggest removing this. Currently it just sounds like a changelog, which doesn't really add any valuable information in my opinion.

**Review by BvB93 — COMMENTED** [review]
Thanks @oscar139, I'd say that this PR is in good shape!
I've got two comments about comments, but things should be good to go afterwards.

**Review by oscar139 — COMMENTED** [review]
Reviewed suggestions

File: numpy/__init__.pyi [hunk 3]
```diff
@@ -2537,11 +2537,22 @@ class object_(generic):
...
+    def day(self) -> int: ...
+    @property
+    def month(self) -> int: ...
+    @property
+    def year(self) -> int: ...
+
+
 class datetime64(generic):
-    @overload
+    @overload
```
Comment (BvB93) [inline numpy/__init__.pyi:2552]:
```suggestion
    @overload
```
Style nit: there are some trailing whitespaces here.

Comment (oscar139) [reply numpy/__init__.pyi:2552]: Thank you kindly for your patience. I will make a greater effort to include matters of style to my contributions. 

I used a remove trailing white space tool and it found one after `# The `datetime64` constructors requires an object with the three attributes below, `. Did you miss this one or are comments ignored?

**Review by BvB93 — APPROVED** [review]
LGTM

**Comment by charris** [comment]
Thanks @oscar139 .

**Pull request merged**